x/bsonx/noreflection_d_encoder.go

```diff
@@ -0,0 +1,862 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package bsonx
+
+import (
+	"fmt"
+	"math"
+	"reflect"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson/bsoncodec"
+	"go.mongodb.org/mongo-driver/bson/bsonrw"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+var (
+	tPrimitiveD          = reflect.TypeOf(primitive.D{})
+	tPrimitiveA          = reflect.TypeOf(primitive.A{})
+	defaultValueEncoders = bsoncodec.DefaultValueEncoders{}
+)
+
+type reflectionFreeDEncoder struct{}
+
+// ReflectionFreeDEncoder is a ValueEncoder for the primitive.D type that does not use reflection.
+var ReflectionFreeDEncoder bsoncodec.ValueEncoder = &reflectionFreeDEncoder{}
+
+func (r *reflectionFreeDEncoder) EncodeValue(ec bsoncodec.EncodeContext, vw bsonrw.ValueWriter, val reflect.Value) error {
+	if !val.IsValid() || val.Type() != tPrimitiveD {
+		return bsoncodec.ValueEncoderError{Name: "DEncodeValue", Types: []reflect.Type{tPrimitiveD}, Received: val}
+	}
+
+	if val.IsNil() {
+		return vw.WriteNull()
+	}
+
+	doc := val.Interface().(primitive.D)
+	return r.encodeDocument(ec, vw, doc)
+}
+
+func (r *reflectionFreeDEncoder) encodeDocumentValue(ec bsoncodec.EncodeContext, vw bsonrw.ValueWriter, v interface{}) error {
```

Comment:
I'm not sure which is odd one, but this signature has the last two parameters flipped compared to, say, the `encodeNumber` function. Can they all be the same?

Reply:
Thanks for catching this. I changed them all to be `(ec, vw, valueToEncode)`.